### PR TITLE
fix: enable werror for dao-api.

### PR DESCRIPTION
**Pull request by jplaisted**
#29 

## Checklist

- [x] The PR conforms to DataHub's [Contributing Guideline](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/linkedin/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [x] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable)

### CODE REVIEW

File: dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java [hunk 1]
```diff
@@ -74,15 +74,15 @@
 
   private static final int DEFAULT_MAX_TRANSACTION_RETRY = 3;
 
-  protected final BaseMetadataEventProducer _producer;
+  protected final BaseMetadataEventProducer<?, ASPECT_UNION, URN> _producer;
```
Comment (keremsahin1) [inline dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java:77]: Any reason why it's not `? extends RecordTemplate` ? This is repeated in multiple lines.

File: dao-api/src/main/java/com/linkedin/metadata/dao/equality/AlwaysFalseEqualityTester.java [hunk 2]
```diff
@@ -1,23 +1,25 @@
...
-import com.linkedin.data.template.DataTemplate;
+import com.linkedin.data.template.RecordTemplate;
 import javax.annotation.Nonnull;
 
 
 /**
  * A {@link EqualityTester} that always returns false.
  */
-public class AlwaysFalseEqualityTester<T extends DataTemplate> implements EqualityTester<T> {
+public class AlwaysFalseEqualityTester<T extends RecordTemplate> implements EqualityTester<T> {
```
Comment (keremsahin1) [inline dao-api/src/main/java/com/linkedin/metadata/dao/equality/AlwaysFalseEqualityTester.java:10]: Why are we switching from DataTemplate to RecordTemplate?

File: dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java (hunk 1 above)
Comment (jplaisted) [reply dao-api/src/main/java/com/linkedin/metadata/dao/BaseLocalDAO.java:77]: `BaseMetadataEventProducer` already has the bounds. No need to respecify it.

File: dao-api/src/main/java/com/linkedin/metadata/dao/equality/AlwaysFalseEqualityTester.java (hunk 2 above)
Comment (jplaisted) [reply dao-api/src/main/java/com/linkedin/metadata/dao/equality/AlwaysFalseEqualityTester.java:10]: DataTemplate was not useful. It is _too_ high level. All the usages (setEqualityTester on BaseLocalDAO) were `RecordTempatae`. I cannot pass a `EqualityTester<?>` to setEqualityTester if the bound is `DataTemplate`, which is basically what happens when you use inference by inlining with `dao.setEqualityTester(AspectFoo.class, AlwaysFalseEqualityTester.instance());`

**Comment by jplaisted** [comment]
ping

File: build.gradle [hunk 3]
```diff
@@ -66,8 +66,10 @@ apply plugin: 'org.shipkit.shipkit-auto-version'
 // than in one big change.
 def wErrorProjects = [
     project(':core-models'),
-    // project(':dao-api'),
-    // project(':dao-api-impl'),
+    project(':dao-api'),
+    // project(':dao-impl:ebean-dao'),
+    // project(':dao-impl:elasticsearch-dao'),
+    // project(':dao-impl:neo4j-dao'),
```
Comment (camelliazhang) [inline build.gradle:72]: What are these commented out lines for?

**Review by camelliazhang — COMMENTED** [review]
Otherwise, LGTM.

File: build.gradle (hunk 3 above)
Comment (jplaisted) [reply build.gradle:72]: See the TODO 4 lines above :)

It can be nice to comment out stuff like this when rolling out; makes it clear what is left.

**Comment by jplaisted** [comment]
Also, since this is a breaking change, and the others will be too, how do we feel about adding them to a separate branch, and then once they're all in merging to master without review? i.e. review small PRs to the other branch, then we can have one big breaking change all at once that was "already reviewed" in chunks?

**Pull request merged**